### ISSUE TITLE
User-facing documentation on how to use our work

### ISSUE THREAD

**Issue by spigo900** [isi-vista/adam#1001]
@marjorief mentioned thinking about things we still need to document for potential open-source users of ADAM.

Assuming these potential users want information on how to extend our work (without changing much of the internals, like how perceptions are generated), I think we still need documentation for the following things. Note that a caret (^) indicates "bonus content" -- things that would be nice to have but seem nontrivial to document and not strictly necessary (I think) for ADAM to be useful.

- Assumptions of our framework?
  - (I think these have already been written up somewhere. I am not sure where they are/whether they are somewhere that potential users would see them.)
- Difference between object learner and object recognizer?
  - (This might already be documented somewhere visible. If not, it's worth pointing out for new users. The purpose of the object recognizer is to experiment with non-object learners under the assumption of perfect object recognition, whether for experiment speed or for isolation-of-faults.)
- Extending the existing framework with
  - a new object.
  - a new property.
  - a new relation.
  - a new action.
  - a new curriculum.
  - ^a new learner.
    - ^Background for implementing new algorithms: the general assumptions our template learner framework makes
        - (Somewhat more detailed than the general assumptions. May also be written up somewhere already.)
        - (I think that a proper tutorial is impossible given there are lots of different algorithms you might want to implement, and implementing e.g. subset is very different from pursuit which is very different from the cross-situational learner. So this would mostly be general assumptions that are implicit in the design so that you can figure out how to adapt your algorithm to our framework. The API documentation should be able to fill in the details.)
    - ^ Implementing new variants of existing learners (ex: pursuit for actions, cross-situational for relations, propose-but-verify for attributes)
      - ^one for each algorithm? subset, pursuit, cross-situational, propose-but-verify
      - (I think it would make sense to include pointers to the different object learner implementations, but not to go into detail about each one.)
  - ^a new language
    - (Would need to consult with @paynesa on this as I haven't worked with the language-specific code enough to be confident I know all the steps in defining a new language.)
    - (This might be messy to document given the determiner hacks and other language-specific stuff spread throughout the codebase. Those things would need to be updated if a user wanted to add a new language.)
- Examples for almost every extension
  - None for adding a completely new learning algorithm
- Running your experiment
  - Changing `log_experiment.py` to include your new curriculum (if applicable)
  - Pointer to Jacob's excellent documentation on the various experiment parameters

I think it should take me a few days at most to produce the documentation without "adding a new learner," without "adding a new language," and without examples. I think it should take a little over a week to produce examples that go along with each non-bonus section, assuming example generation goes poorly. I expect adding an object, adding a new curriculum, and "running your experiment" to be the simplest three to generate examples for, while (of the non-bonus content) adding a property, relation or action might be more complicated.

**Comment by spigo900** [comment]
I'm thinking it makes the most sense to structure the "extending/examples/experiments" documentation as one document. The first part would be a high-level overview of each point, and the second would be a step-by-step example of how to put all those ideas together. The example would not be a particularly interesting experiment, but would demonstrate how to set up an experiment.

Re: the example, here's my idea. We want to do an experiment involving a snake toy for some reason. We want to add a toy snake object with the property that it's flexible. We want to add a relation "wrapped around." We want to add an action "pick up." The experiment would involve showing the toy alone, then the toy wrapped around a (toy) truck, say, then a baby picking up the toy. I think this should be good enough for demonstration purposes.

**Comment by spigo900** [comment]
It looks like adding this example relation is more complicated than I thought and would be out of scope for this document. It looks to me like adding this new relation would require either (1) modifying our `Region` class to handle a region like "wrapped around," which would require changing a lot of things, (2) interpreting distance flexibly (so that `WRAPPED_AROUND` is a distance), which might break things, or (3) doing a hack where we represent "wrapped around" by adding an "exterior but in contact with" region along every (non-primary) axis of the wrapped object. I think (3) is probably the best solution but it's complicated enough that it might get in the way of the demo.

If anyone has an idea for a simpler relation, or thoughts on this, those are welcome.

**Comment by spigo900** [comment]
I think it would also make sense to add a template parameters file for `log_experiment.py` to make it simpler to create new experiments.

**Comment by spigo900** [comment]
I originally structured this as one document, but I now think it makes more sense to structure this as several smaller documents:

- Using ADAM: links to tutorials, in order
  1. Running experiments
  2. Setting up a new curriculum
      - plus running an experiment that uses this new curriculum
  3. Extending ADAM with new
     - objects
     - properties
     - ...

(with probably one document per extension type.)

"Using ADAM" would be an overview document explaining what this documentation is -- a brief introduction regarding how to use ADAM to run experiments and extend it in simple ways. It's the "getting started" guide to complement the API reference's less connected "everything you might possibly want to know" documentation. Each smaller document would be fairly short and focus on a specific task, with a high level overview followed by a concrete example.

So, for example, section 1 would tell you about experiments. It would explain that experiments are (mostly) run using `log_experiment.py` and are defined by parameters files. It would briefly discuss parameters files, referring to Jacob's documentation for more detail. Finally, it would walk the reader through a short example of how to run an experiment:

1. Set up your root parameters file.
   1. A reminder in case they missed that step in the README, because this "set up `root.params`" step is sort of odd and specific to ISI VISTA projects.
2. Copy the template and change two lines.
3. Run the script.

(I am somewhat less sure about this particular example than the others, but it seems worth including.)

**Comment by spigo900** [comment]
We now have documentation on

1. running experiments,
2. creating simple curricula,
3. adding objects, and
4. adding binary properties and colors.

I think actions should come next, however I am not sure what should come after that. Going by the original outline, I still need to write documentation on both adding actions and adding relations. I think I can give a step by step tutorial for actions that is useful enough for actions that fit into our framework. However, adding relations seems hard, significantly less "step-by-step" than either objects or properties.

Specifically, adding nontrivial relations seems either hard to do at all, or likely to either fail or break things without code changes. I'm not sure if we even have a generic way to create new edge-style relations without modifying the internals, specifically the perception generation code. I might be wrong though. Even then, though, we don't allow multiple edges between nodes, and we don't (I think) support reifying relations as node. That means if you try to add a custom relation where there's already an implicit edge, it either won't work or might break something else. Overall it seems like a tutorial would be quite fragile and not very useful.

It might be worth including some documentation on relations anyway, but not a tutorial on how to add one -- more of a summary of what parts of the code handle relations and what existing code/helpers we have for defining relations.

It might help to have a little more information on curricula (for example, the helpers; including actions in a template; the various relation arguments (asserted_always, before_action, after_action, constraining_relations) though I'm not sure how to organize it.

**Comment by LichMaster98** [comment]
@spigo900 With regards to relations defining a new one is more complicated than adding an attribute to the library if an implementer wants utility functions to automatically create the appropriate `Region` relations (recommended method). Somewhere we should explain how the utility methods which create these relations works and how to create new functions to extend the set of functions available. 

While I agree that future users could hit problems adding new relations with a more complicated structure. You mentioned “wrapped around” in one of your comments. With our current structure I’d say that’s `exterior-but-in-contact` between the two objects and then there would need to be some axis alignment information between the two objects. The biggest difficulty is the axis information here. If someone wants to implement more robust region descriptions then implementing some form of multidigraph (hypergraph) versions of our core functions (matching & relaxation) is probably necessary. Alternatively the ability to turn a relation into a more complete set of node is the recommendation to continue development in that area. 

With regards to curriculum creation I’d recommend walking through what the template `Phase1SituationTemplate` (which is poorly named as I think the ‘phase1’ monicker here isn’t accurate) can accept as you mentioned eg: `salient_object_variables` & `asserted_always_relations`. As those are certainly the confusing section of creating a new template. 

Secondly on curriculum creation some explanation of the layered approach to how the sequence of the final curriculum is generated from templates (sampled vs all). Additionally if one is going to use the pre-generated curriculum then we need to probably add a utility/option to turn the curriculum from a generator into an explicit list of instances. The latter takes more RAM during experiment run (as the entire curriculum is generated in advance) but that means the experiment thread no longer has to do the work of generating all the perception graphs as its done prior. (I discovered this quirk in the integrated learner example and switching to what we probably intended by the curriculum pre-generation took me from ~200 examples in 12 hours to nearly ~500 examples during the experiment while only taking pregeneration time of ~1 hour for a total of close to 1200 instances)

**Comment by spigo900** [comment]
@LichMaster98 Region relations do seem like the best option for a "how to add new relations" tutorial. So, to clarify, your suggestion is to explain how the existing region relations work and how to use the `make_*_dsl_region_relation` functions to create new region relation functions? Unless you meant creating new `make_*_dsl_region_relation` functions.

I think I agree about wrapped around. That seems complicated enough to be out of scope for this documentation.

I agree it would be useful to walk through working with situation templates. I've written some materials on this as part of adding a simple curriculum, and I think it makes sense to go into a little more detail in a separate "advanced curriculum" tutorial.

I am not quite sure what you mean about explaining the layered approach. Explain how to use the functions (sampled/all) to turn templates into situations, then generate perceptions/language from them? Explain the difference between `sampled` and `all`?

Re: pre-generation, I'm not sure what you mean. I think we're talking about adding an option to `log_experiment.py`, but are we about the option being:
1. to list-ify the curriculum before using it, or
2. to load the curriculum if available, otherwise list-ify and serialize the curriculum before using it?

Thanks for the feedback!

**Comment by LichMaster98** [comment]
Correct @spigo900, just explaining how the `make_*_dsl_region_relation` function works is my intent. 

So when writing a curriculum I generally divide the task into two parts the first is creating all the situation templates needed for the curriculum. The second is selecting the type of objects to fill the requisite slots in the templates and any properties required by those objects. That template then gets passed into a method which takes the templates and turns the template information into a real situation. The functions of `sampled` Or `all` (forget the actual function name at this moment) both do the task of turning a template into a real situation but sampled only produces a maximum number of examples (parameter) while all tries to create all situations which fit the requested template. For more complex templates, especially those including noise, using the all function can produce an unreadably large number of situations so its something to be aware of as a curriculum creator. 

With regards to curriculum pre-generation an option to list-ify would need to be added to the script which writes the curriculum to a pickle file. Understandably our standard methods to generate instances generates the situation and linguistic utterance however it leaves creating the perceptual representation until called by the experiment runner (the iterable of the instances is a generator) so this means that the experiment thread has to do the work of generating the initial perception graph for the instance. Therefore if multiple experiments are using the same curriculum they are all creating these perception graphs multiple times rather than making them when the curriculum was initially created and saving them into the pickle file. Does that make more sense?

**Comment by spigo900** [comment]
@LichMaster98 Thank you for the confirmation.

Right, I agree about creating a template. So I guess the two things I should be careful to note here when I document this are:
1. When you create templates using actions, you need to be careful to choose bounds on your object variables such that they can actually fill the slots you want them to. (Let me know if I misunderstood your second step.)
2. `all_possible` can create a lot of situations, way more than you probably want, so be careful with it.
Thanks for mentioning these -- I hadn't thought of those complications.

Re: pre-generation, I think I am on the same page now. So if I understand right, the problem *isn't* that we don't list-ify our instances/generate perceptual representations at all when pre-generating the curriculum. Rather, the problem is that we don't pre-generate the *perception graphs*, and this ends up being most/a lot of the work.

It seems like it might worthwhile to change this, but I am not sure how it would work. You said you implemented this; how did you do it?

**Comment by LichMaster98** [comment]
@spigo900 So I implemented this for the curriculum used in the integrated learners experiment (and by accident in the language ablation experiment though I now know I did it) by using the [Explicit Instance Group](https://github.com/isi-vista/adam/blob/674bb6ca6e9d780eac5d27a66c456f8419c8f7ea/adam/curriculum/__init__.py#L81-L109) class to hold the instances. 

There are some trade-offs to doing this though. First, any system running an experiment with the entire curriculum generated up front will require additional RAM as the entire curriculum is then held in RAM rather than in the generator expression. (It’s still minor in the grand scheme of things the experiment setups for the integrated learners are requesting ~16 GB and I don’t believe I use anywhere near all of it but 8GB was too little). There’s also longer spent generating the curriculum (via the perception graphs) before the experiments start. 

For single experiment runs the generator version is more efficient for rapid development testing if you need to modify the curriculum.

**Comment by spigo900** [comment]
@LichMaster98 So it sounds like the idea is to have an option where we convert our curriculum (an iterable of generated instance groups) to a *list* of *explicit* instance groups before running experiments on it in `log_experiment`. That is, this option is unrelated to the script that pre-generates and pickles a curriculum (`generate_curriculum`).

We should already convert to explicit instance groups in [pre-generation](https://github.com/isi-vista/adam/blob/674bb6ca6e9d780eac5d27a66c456f8419c8f7ea/adam/experiment/generate_curriculum.py#L38), unless there's a bug.

**Comment by LichMaster98** [comment]
Ah @spigo900 in that case maybe just mention the potential RAM impacts on larger curriculums when using pre-generation. (I’ll assume for the moment then that my performance improvements came as a result of fixing the bugs around my shuffle method rather than the curriculum export script then)

**Comment by spigo900** [comment]
@LichMaster98 Sounds good!

**Comment by spigo900** [comment]
We may want to change our API documentation formatting slightly:

![image](https://user-images.githubusercontent.com/6877173/102518348-2c0c3380-405f-11eb-8a12-eccbec107c3d.png)

Or not, since this only seems to affect the two `instances` functions.

Each of the argument lists goes on for ~2.5 pages, mostly because of the default values and their ontologies.

**Comment by spigo900** [comment]
We now have (in the PR) tutorials on:

1. Running experiments (ours and your own),
2. Defining simple ("one object") curricula,
3. Defining more complex curricula (with relations and objects),
4. Adding objects,
5. Adding properties,
6. Adding actions, and
7. Adding relations.

I've also made some minor quality-of-life changes to the existing documentation, mostly just fixing out-of-date type references (`LanguageLearner` instead of `TopLevelLanguageLearner`, for example).

As for next steps, the immediate one is to test the instructions/code I gave in each of these tutorials and confirm that they actually work; it wouldn't be very useful if it turned out the code wasn't enough, had bugs, etc.

If I have time, I may also add a brief tutorial on how to pre-generate the curriculum for use in experiments (with a warning about RAM as @LichMaster98 suggested).

If anyone wants to review the documentation, I think it's in a final-ish enough state now to do that.
